### DISCUSSION TITLE
Basic Authentication - Request.Headers["Authorization"] is always null

### DISCUSSION THREAD

**Discussion by enricofacchinetti** [CoreWCF/CoreWCF#827] [General]
Hi,
here is our code implementing basic authentication in CoreWCF.
We are aware of the protocol's security issues and in the next future we will replace it; but in the short term we have to meet some backward compatibility requirements.
In the Program.cs file, Just before to invoke the method app.UseServiceModel() we implemented a middleware to launch the basic authentication. 

The method HandleAuthenticateAsync() is then invoked in a custom class inheriting AuthenticationHandler.
But in this method, when it's up to fetch the Authorization header invoking Request.Headers["Authorization"], a null object is returned.
We would like to point out that the other headers are present in the IHeaderDictionary Request.Headers.

We tried to set the BasicHttpSecurityMode to Transport and to TransportWithMessageCredential.
But in both the cases, the result doesn't change.
The client we use to call our CoreWCF service during our tests is SoapUI 5.7.0.
We set the Authorization section of SoapUI 5.7.0. to Basic and we set the credentials.
We tried to find something relevant in the documentation and in the forum but without success.
Does anybody has any ideas on how to fix this problem?
Thanks in advance for your precious help.
E.

// Program.cs

`
var builder = WebApplication.CreateBuilder();

builder.Configuration.AddJsonFile("libsettings.json", optional: true, reloadOnChange: true);
builder.Configuration.AddJsonFile("appsettings.json", optional: true, reloadOnChange: true);
builder.Configuration.AddEnvironmentVariables();

builder.WebHost.ConfigureKestrel((context, options) =>
{
     options.AllowSynchronousIO = true;
});

// Add WSDL support
builder.Services.AddServiceModelServices();
builder.Services.AddServiceModelMetadata();
builder.Services.AddSingleton<IServiceBehavior, UseRequestHeadersForMetadataAddressBehavior>();

builder.Services.AddTransient((provider) =>
{
     var libSettings = builder.Configuration.GetSection("LibSettings").Get<LibSettings>();
     var connString = builder.Configuration.GetConnectionString("MyConnString");
     var contextOptions = new DbContextOptionsBuilder<TransApiContext>()
                            .UseSqlServer(connString)
                            .Options;
     return new MySoapServiceClass(libSettings, contextOptions);
});

builder.Services.AddControllers();

if (ServicePointManager.SecurityProtocol.HasFlag(SecurityProtocolType.Tls12) == false)
{
     ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12 | SecurityProtocolType.Tls11 | SecurityProtocolType.Tls;
}

builder.Services.AddAuthentication("BasicAuthentication").
            AddScheme<AuthenticationSchemeOptions, BasicAuthenticationHandler>
            ("BasicAuthentication", null);

var app = builder.Build();

app.Use(async (context, next) =>
{
     // Only check for basic auth when path is for the TransportWithMessageCredential endpoint only
     if (context.Request.Path.StartsWithSegments("/myBaseAddress/myservice"))
     {
         // Check if currently authenticated
         var authResult = await context.AuthenticateAsync("BasicAuthentication");
         if (authResult.None)
         {
             // If the client hasn't authenticated, send a challenge to the client and complete request
             await context.ChallengeAsync("BasicAuthentication");
             return;
         }
     }
     // Call the next delegate/middleware in the pipeline.
     await next(context);
});

app.UseServiceModel(builder =>
{
     builder.AddService<PayPerUse>(serviceOptions =>
     {
         serviceOptions.DebugBehavior.IncludeExceptionDetailInFaults = true;
         serviceOptions.BaseAddresses.Add(new Uri("http://localhost/myBaseAddress"));
         serviceOptions.BaseAddresses.Add(new Uri("https://localhost/myBaseAddress"));
         serviceOptions.DebugBehavior.HttpsHelpPageEnabled = true;
         serviceOptions.DebugBehavior.HttpsHelpPageUrl = new Uri("https://localhost/myBaseAddress/help");
         serviceOptions.DebugBehavior.HttpHelpPageUrl = new Uri("http://localhost/myBaseAddress/help");
     })
     .AddServiceEndpoint<MySoapServiceClass, IMySoapServiceClass>(new BasicHttpBinding(BasicHttpSecurityMode.Transport), "/myservice");
     var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
     serviceMetadataBehavior.HttpsGetEnabled = true;
});

app.UseHttpsRedirection();
app.UseRouting();
app.UseAuthentication();
app.UseAuthorization();
app.UseEndpoints(endpoints =>
{
     endpoints.MapControllers();
});

app.Run();

   public class BasicAuthenticationHandler : AuthenticationHandler<AuthenticationSchemeOptions>
    {
        public BasicAuthenticationHandler(IOptionsMonitor<AuthenticationSchemeOptions> options, ILoggerFactory logger,
                                          UrlEncoder encoder, ISystemClock clock) : base(options, logger, encoder, clock)
        {
        }

        protected override Task<AuthenticateResult> HandleAuthenticateAsync()
        {
            try
            {
                var auth = (string)Request.Headers["Authorization"];

                // Authorization header is always null but the client sends it 
                if (string.IsNullOrEmpty(auth)) return Task.FromResult(AuthenticateResult.Fail("Invalid Credentials"));
            }
            catch
            {
                return Task.FromResult(AuthenticateResult.Fail("Error Occured.Authorization failed."));
            }
        }
    }
`

**Comment by toreaurstadboss** [comment]
I have tested also Basic Authentication using Core.WCF. I have followed the tips shown in different threads and googling around and I have not found by username and password. The problem is that if I try to look into the Request or via HttpContextAccessor to look for username and password inside the AuthenticationHandler, I get an error that "Response ended prematurely". 

A simple Service is added in this repo, using BasicAuth : 
https://github.com/toreaurstadboss/CoreWCFService1

Here is a Client that accesss this service in this repo: 
https://github.com/toreaurstadboss/CoreWCFWebClient1

The client request to service is inside Index.cshtml of the client

Request does not contain the username or password sent when retrieved on the serverside. 

If I try to send the username and password via headers like this : 

```
    string username = "someuser"; 
    string password = "somepassw0rd"; 

    string basicHeaderValue = "Basic " + Base64Encode($"{username}:{password}");

    var eab = new EndpointAddressBuilder(new EndpointAddress("https://localhost:7218/Service.svc"));

    eab.Headers.Add(AddressHeader.CreateAddressHeader("Authorization",  // Header Name
        string.Empty,           // Namespace
        basicHeaderValue));  // Header Value

    //eab.Headers.Add(AddressHeader.CreateAddressHeader("Connection",  // Header Name
    //    string.Empty,           // Namespace
    //    "Keep-Alive"));  // Header Value

    var endpointAddress = eab.ToEndpointAddress();

    var client = new ServiceClient(binding, endpointAddress);

    client.ClientCredentials.UserName.UserName = username;
    client.ClientCredentials.UserName.Password = password;

    var result =  client.GetDataAsync(42).Result;
```
The body of my request looks like this : 

```
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Header><Authorization>dXNlcjE6cGFzc3cwcmQ=</Authorization></s:Header><s:Body><GetData xmlns="http://tempuri.org/"><value>42</value></GetData></s:Body></s:Envelope>
```
The Authorization header is available in the Request.Body using this approach.

The AuthenticationHandler looks like this: 

```
public class BasicAuthenticationHandler : AuthenticationHandler<AuthenticationSchemeOptions>
{

    private readonly IUserRepository _userRepository;
    public BasicAuthenticationHandler(IOptionsMonitor<AuthenticationSchemeOptions> options,
        ILoggerFactory logger,
        UrlEncoder encoder,
        ISystemClock clock, IUserRepository userRepository) :
       base(options, logger, encoder, clock)
    {
        _userRepository = userRepository;
    }

    protected async override Task<AuthenticateResult> HandleAuthenticateAsync()
    {
        string? authTicketFromSoapEnvelope = Request.GetAuthenticationHeaderFromSoapEnvelope();

        if (authTicketFromSoapEnvelope != null && authTicketFromSoapEnvelope.StartsWith("basic", StringComparison.OrdinalIgnoreCase))
        {
            var token = authTicketFromSoapEnvelope.Substring("Basic ".Length).Trim();
            var credentialsAsEncodedString = Encoding.UTF8.GetString(Convert.FromBase64String(token));
            var credentials = credentialsAsEncodedString.Split(':');
            if (await _userRepository.Authenticate(credentials[0], credentials[1]))
            {
                //var claims = new[] { new Claim("name", credentials[0]), new Claim(ClaimTypes.Role, "Admin") };
                var identity = new GenericIdentity(credentials[0]);
                var claimsPrincipal = new ClaimsPrincipal(identity);
                var ticket = new AuthenticationTicket(claimsPrincipal, Scheme.Name);
                return await Task.FromResult(AuthenticateResult.Success(ticket));
            }
        }

        Response.StatusCode = 401;
        Response.Headers.Add("WWW-Authenticate", "Basic realm=\"thoushaltnotpass.com\"");
        return await Task.FromResult(AuthenticateResult.Fail("Invalid Authorization Header"));
    }

```
The UserRepository looks like this: 

```

 public interface IUserRepository
    {

        public Task<bool> Authenticate(string username, string password);
    }

    public class UserRepository : IUserRepository
    {
        public Task<bool> Authenticate(string username, string password)
        {
            //TODO: some dummie auth mechanism used here, make something more realistic such as DB user repo lookup or similar
            if (username == "someuser" && password == "somepassw0rd")
            {
                return Task.FromResult(true);
            }
            return Task.FromResult(false);
        }
    }
```
I managed to authenticate the client call using this, but I get the error "The response ended prematurely". 

As of now, it looks like many developers cannot make Basic Auth work with WCF and I tried a lot of different approaches. It would be nice if you could clone the two repos I mentioned above and pinpoint me in which place I have failed setting it up. Please note, I have worked some years with WCF, but I am relatively novice with CoreWCF.

**Reply by toreaurstadboss** [reply]
I have updated my two repos which I mentioned and managed to send the client credentials instead via headers in the soap envelope.  Reading those headers on the serverside took a lot of trial and error to avoid ASP.NET Core bugs. 

This method finally worked : 

The important bit is the use of the AdvanceTo method and using BodyReader. It is the only way I found out to avoid having ASP.NET errors not crash the request and response with the "The response ended prematurely". 

But it seems like CoreWCF cannot transmit the ClientCredentials properly but instead have to make use of the EndpointAddressBuilder.

I'll tidy up my two repos and put a blog post on my blog post "Coding grounds" tomorrow or next day, by the way covering what I found out. Hopefully CoreWCF comes with some good mechanism for doing BasicAuth with added helper methods for this in future releases. 

By using GenericPrincipal I managed to set the username too in OperationContext. 

```
 public static async Task<string>? GetAuthenticationHeaderFromSoapEnvelope(this HttpRequest request)
    {
       
            ReadResult requestBodyInBytes = await request.BodyReader.ReadAsync();
        //request.BodyReader.AdvanceTo(requestBodyInBytes.Buffer.Start, requestBodyInBytes.Buffer.End);

        string body = Encoding.UTF8.GetString(requestBodyInBytes.Buffer.FirstSpan);
            request.BodyReader.AdvanceTo(requestBodyInBytes.Buffer.Start, requestBodyInBytes.Buffer.End);

            //request.Body.CopyTo(streamCopy);
            //streamCopy.Position = 0; // rewind

            //string body = new StreamReader(streamCopy).ReadToEnd();

            string authTicketFromHeader = null;

            if (body?.Contains(@"http://schemas.xmlsoap.org/soap/envelope/") == true)
            {
                XNamespace ns = "http://schemas.xmlsoap.org/soap/envelope/";
                var soapEnvelope = XDocument.Parse(body);
                var headers = soapEnvelope.Descendants(ns + "Header").ToList();

                foreach (var header in headers)
                {
                    var authorizationElement = header.Element("Authorization");
                    if (!string.IsNullOrWhiteSpace(authorizationElement?.Value))
                    {
                        authTicketFromHeader = authorizationElement.Value;
                        break;
                    }
                }
            }

            //streamCopy.Position = 0; // rewind again
            //request.Body = streamCopy; // put back in place for downstream handlers

            //request.Headers.Add("Authorization", authTicketFromHeader);

            return authTicketFromHeader;

    }
```
